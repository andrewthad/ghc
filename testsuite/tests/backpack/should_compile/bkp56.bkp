{-# language DataKinds #-}
{-# language GADTs #-}
{-# language KindSignatures #-}
{-# language TypeFamilies #-}
{-# language UnsaturatedTypeFamilies #-}

unit i where
    module I where
        import Data.Kind (Type)
        data SingBool :: Bool -> Type where
          SingTrue :: SingBool 'True
          SingFalse :: SingBool 'False
        data FlipFlop :: Bool -> Type where
          Flip :: Int -> FlipFlop 'True
          Flop :: Char -> FlipFlop 'False

unit p where
    dependency i
    signature H where
        import I (FlipFlop)
        data Foo :: Bool ~> Bool
        x :: FlipFlop (Foo 'False)
        y :: FlipFlop (Foo 'True)
    module A where
        import H
        import I
        choose :: SingBool b -> FlipFlop (Foo b)
        choose SingFalse = x
        choose SingTrue = y

unit h where
    dependency i
    module H where
        import I
        type family Foo (b :: Bool) :: Bool where
          Foo 'True = 'False
          Foo 'False = 'True
        x :: FlipFlop 'True
        x = Flip 5
        y :: FlipFlop 'False
        y = Flop 'a'

unit r where
    dependency h
    dependency p[H=h:H]
