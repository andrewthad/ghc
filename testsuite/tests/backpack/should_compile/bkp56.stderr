Fill this in once compilation works
